crates/re_viewport/src/system_execution.rs

```diff
@@ -121,16 +80,68 @@ pub fn execute_systems_for_space_view<'a>(
     }
 
     let query = re_viewer_context::ViewQuery {
-        space_view_id: space_view.id,
-        space_origin: &space_view.space_origin,
+        space_view_id: view.id,
+        space_origin: &view.space_origin,
         per_visualizer_data_results,
         timeline: *ctx.rec_cfg.time_ctrl.read().timeline(),
         latest_at,
         highlights,
     };
 
-    let system_output =
-        create_and_run_space_view_systems(ctx, *space_view.class_identifier(), &query);
+    let mut context_systems = ctx
+        .space_view_class_registry
+        .new_context_collection(view.class_identifier());
+    let mut view_systems = ctx
+        .space_view_class_registry
+        .new_visualizer_collection(view.class_identifier());
+
+    let draw_data = if let Some(view_state) = view_states.get(view.id) {
+        run_space_view_systems(
+            ctx,
+            view.class_identifier(),
+            &query,
+            view_state.view_state.as_ref(),
+            &mut context_systems,
+            &mut view_systems,
+        )
+    } else {
+        re_log::error_once!("No view state found for view {}", view.id);
+        Vec::new()
+    };
+
+    (
+        query,
+        SystemExecutionOutput {
+            view_systems,
+            context_systems,
+            draw_data,
+        },
+    )
+}
 
-    (query, system_output)
+pub fn execute_systems_for_all_views<'a>(
+    ctx: &'a ViewerContext<'a>,
+    tree: &egui_tiles::Tree<SpaceViewId>,
+    views: &'a BTreeMap<SpaceViewId, SpaceViewBlueprint>,
+    view_states: &ViewStates,
+) -> HashMap<SpaceViewId, (ViewQuery<'a>, SystemExecutionOutput)> {
+    let Some(time_int) = ctx.rec_cfg.time_ctrl.read().time_int() else {
+        return HashMap::default();
```

Comment:
nit: less diff later on if we change the return type to e.g. a BTtreeMap

```suggestion
        return Default::default();
```